### PR TITLE
Avoid capturing raw `this` pointer in WebSocket module

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description
- Defines `WebSocketModule::SharedState` struct.
- Enforces weak-pointer-only captures within `WebSocketModule`.
- Removes URL as a member of `IWebSocketResource` implementations.

### Type of Change
- Bug fix

### Why
CxxModules frequently pass their resources to asynchronous tasks by capturing the raw `this` pointer, which can lead to invalid memory access attempts due to race conditions between the thread owning the module and threads executing the asynchronous work defined in `getMethods()`.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9247)

### CODE REVIEW

File: vnext/Shared/Modules/WebSocketModule.cpp [hunk 1]
```diff
@@ -50,7 +56,7 @@ std::vector<facebook::xplat::module::CxxModule::Method> WebSocketModule::getMeth
 {
   return
   {
-    Method(
+    {
       "connect",
       [this](dynamic args) // const string& url, dynamic protocols, dynamic options, int64_t id
```
Comment (NikoAri) [inline vnext/Shared/Modules/WebSocketModule.cpp:61]: >this [](http://example.com/codeflow?start=7&length=4)

Don't you need to pass weak m_holder here?

Comment (JunielKatarn) [reply vnext/Shared/Modules/WebSocketModule.cpp:61]: Possibly. This is work in progress. `WebSocketModule` will require some refactoring.

File: vnext/Desktop.UnitTests/BaseWebSocketTests.cpp [hunk 2]
```diff
@@ -65,7 +65,7 @@ TEST_CLASS(BaseWebSocketTest){
       return make_error_code(errc::state_not_recoverable);
     });
 
-    ws->Connect({}, {});
+    ws->Connect("ws://localhost", {}, {});
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/BaseWebSocketTests.cpp:68]: >"ws://localhost" [](http://example.com/codeflow?start=16&length=16)

(nit) can you make it a shared string const, and just reference it in all these places

File: vnext/Desktop.UnitTests/WinRTWebSocketResourceUnitTest.cpp [hunk 3]
```diff
@@ -77,12 +76,11 @@ TEST_CLASS (WinRTWebSocketResourceUnitTest) {
...
 
     // Test APIs
-    auto rc =
-        make_shared<WinRTWebSocketResource>(std::move(imws), MockDataWriter{}, Uri{L"ws://host:0"}, CertExceptions{});
+    auto rc = make_shared<WinRTWebSocketResource>(std::move(imws), MockDataWriter{}, CertExceptions{});
     rc->SetOnConnect([&connected]() { connected = true; });
     rc->SetOnError([&errorMessage](Error &&error) { errorMessage = error.Message; });
 
-    rc->Connect({}, {});
+    rc->Connect("ws://host:0", {}, {});
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/WinRTWebSocketResourceUnitTest.cpp:83]: >"ws: [](http://example.com/codeflow?start=16&length=4)

(nit) same, make a shared const char* for this string

File: vnext/Shared/WinRTWebSocketResource.cpp [hunk 4]
```diff
@@ -123,13 +117,14 @@ WinRTWebSocketResource::~WinRTWebSocketResource() noexcept /*override*/
 
 #pragma region Private members
 
-IAsyncAction WinRTWebSocketResource::PerformConnect() noexcept {
+IAsyncAction WinRTWebSocketResource::PerformConnect(Uri &&uri) noexcept {
   auto self = shared_from_this();
+  auto coUri = std::move(uri);
```
Comment (NikoAri) [inline vnext/Shared/WinRTWebSocketResource.cpp:122]: >coUri [](http://example.com/codeflow?start=7&length=5)

Why is this needed? Why not just do std::move on line 127?

Comment (NikoAri) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: Oh, because we want to std::move before co_await?

Comment (NikoAri) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: But still, why do we need that here? wouldn't "uri" be still available after co_await?

File: vnext/Shared/WinRTWebSocketResource.cpp [hunk 5]
```diff
@@ -348,7 +343,24 @@ void WinRTWebSocketResource::Connect(const Protocols &protocols, const Options &
...
 
   m_connectRequested = true;
-  PerformConnect();
+
+  Uri uri{nullptr};
+  try {
+    uri = Uri{winrt::to_hstring(std::move(url))};
+  } catch (hresult_error const &e) {
+    if (m_errorHandler) {
+      m_errorHandler({HResultToString(e), ErrorType::Connection});
```
Comment (NikoAri) [inline vnext/Shared/WinRTWebSocketResource.cpp:352]: >HResultToString [](http://example.com/codeflow?start=22&length=15)

(nit) although converting to string is nice here, it would still be good to propagate the actual HResult value into m_errorHandler as well.

File: vnext/Shared/Modules/WebSocketModule.cpp [hunk 6]
```diff
@@ -75,166 +171,83 @@ std::vector<facebook::xplat::module::CxxModule::Method> WebSocketModule::getMeth
           }
         }
 
-        weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 3), jsArgAsString(args, 0));
+        weak_ptr weakWs = GetOrCreateWebSocket(jsArgAsInt(args, 3), jsArgAsString(args, 0), weakState);
```
Comment (NikoAri) [inline vnext/Shared/Modules/WebSocketModule.cpp:174]: >GetOrCreateWebSocket [](http://example.com/codeflow?start=26&length=20)

Hmm, so why not just do:
strongState = weakState.lock
if (!strongState) return
strongState->GetOrCreateWebSocket(...)
?

**Review by NikoAri — APPROVED** [review]
:shipit:

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: Nope. Unless it's a thread-safe pass by reference (it's not), a new lvalue must be held for the co-routine to keep it when it resumes.

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 5 above)
Comment (JunielKatarn) [reply vnext/Shared/WinRTWebSocketResource.cpp:352]: This was discussed before.

Keeping the HRESULT would be both overkill because only the error string needs to be propagated to the JS layer , and this would potentially break the abstraction (IWebSocketResource is CppWinRT and COM-agnostic).

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: See https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency#parameter-passing

File: vnext/Shared/Modules/WebSocketModule.cpp (hunk 6 above)
Comment (JunielKatarn) [reply vnext/Shared/Modules/WebSocketModule.cpp:174]: This change removes `GetOrCreateWebSocket` as a member of `WebSocketModule` altogether and turns it into a static free function.

Same applies to `SendEvent`.

This was done to decouple the functionality that is not inherent to `CxxModule` and its lifetime.

**Review by NikoAri — APPROVED** [review]

File: vnext/Desktop/BeastWebSocketResource.cpp [hunk 7]
```diff
@@ -30,6 +30,90 @@ using std::unique_ptr;
...
+
+    m_concreteResource = std::make_shared<Beast::WebSocketResource>(std::move(url));
+  } else if (url.scheme == "wss") {
+    if (url.port.empty())
+      url.port = "443";
+
+    m_concreteResource = std::make_shared<Beast::SecureWebSocketResource>(std::move(url));
+  }
+
+  m_concreteResource->SetOnClose(std::move(m_closeHandler));
```
Comment (vmoroz) [inline vnext/Desktop/BeastWebSocketResource.cpp:57]: >std::move [](http://example.com/codeflow?start=33&length=9)

If we use std::move here, then do we expect that the fields are becoming invalid after the Connect call? I.e. we never expect to call Connect the second time.

**Comment by vmoroz** [comment]
>   m_stream->set_option(websocket::stream_base::decorator([options = std::move(options)](websocket::request_type &req) {

options are const& - we cannot move them

---
Refers to: vnext/Desktop/BeastWebSocketResource.cpp:410 in 3564788. [](commit_id = 3564788d8de0c8ee66d79a6ef5d140ad1bd05891, deletion_comment = False)

**Comment by vmoroz** [comment]
>   m_connectHandler = handler;

The handler is passed as an r-value reference, but to use the move-semantic we must use the std::move() here and in the methods below.

---
Refers to: vnext/Desktop/BeastWebSocketResource.cpp:523 in 3564788. [](commit_id = 3564788d8de0c8ee66d79a6ef5d140ad1bd05891, deletion_comment = False)

File: vnext/Desktop/BeastWebSocketResource.cpp (hunk 7 above)
Comment (JunielKatarn) [reply vnext/Desktop/BeastWebSocketResource.cpp:57]: Correct!

Also note, this is the legacy (Boost/Beast) implementation which I keep around for experimentation and performance comparison.

It's by default not even built into the RNW DLL, and will eventually be deleted.
Only keeping it up to date with the `IWebSocket` semantics.

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 4 above)
Comment (vmoroz) [inline vnext/Shared/WinRTWebSocketResource.cpp:122]: >= std::move(uri); [](http://example.com/codeflow?start=13&length=17)

I wonder what the result of this operation is. Are you sure that the Uri constructor is called and we not keep just a reference?
I would rather be explicit. E.g.
Uri coUri{std::move(uri)};

File: vnext/Shared/WinRTWebSocketResource.cpp [hunk 8]
```diff
@@ -348,7 +343,24 @@ void WinRTWebSocketResource::Connect(const Protocols &protocols, const Options &
   }
 
   m_connectRequested = true;
-  PerformConnect();
+
+  Uri uri{nullptr};
+  try {
+    uri = Uri{winrt::to_hstring(std::move(url))};
```
Comment (vmoroz) [inline vnext/Shared/WinRTWebSocketResource.cpp:349]: >std::move( [](http://example.com/codeflow?start=32&length=10)

not needed: the to_hstring cannot move the input inside.

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: I can triple-check, but not keeping this reference made the `uri` inaccessible by the coroutine when it resumed.
In short, it seemed like the value was not being passed to the resuming thread.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Shared/WinRTWebSocketResource.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/WinRTWebSocketResource.cpp:122]: Verified. Not keeping a new reference crashes tests.
